crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1108,19 +1126,75 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_set_comprehension_expression(&mut self, setcomp: &ast::ExprSetComp) -> Type<'db> {
         let ast::ExprSetComp {
             range: _,
-            elt,
+            elt: _,
             generators,
         } = setcomp;
-        self.infer_expression(elt);
-        for generator in generators {
-            self.infer_comprehension(generator);
-        }
+
+        self.infer_first_comprehension_iter(generators);
 
         // TODO set type
         Type::Unknown
     }
 
-    fn infer_comprehension(&mut self, comprehension: &ast::Comprehension) -> Type<'db> {
+    fn infer_generator_expression_in_scope(&mut self, generator: &ast::ExprGenerator) {
+        let ast::ExprGenerator {
+            range: _,
+            elt,
+            generators,
+            parenthesized: _,
+        } = generator;
+
+        self.infer_comprehensions(generators);
+        self.infer_expression(elt);
+    }
+
+    fn infer_list_comprehension_expression_in_scope(&mut self, listcomp: &ast::ExprListComp) {
+        let ast::ExprListComp {
+            range: _,
+            elt,
+            generators,
+        } = listcomp;
+
+        self.infer_comprehensions(generators);
+        self.infer_expression(elt);
+    }
+
+    fn infer_dict_comprehension_expression_in_scope(&mut self, dictcomp: &ast::ExprDictComp) {
+        let ast::ExprDictComp {
+            range: _,
+            key,
+            value,
+            generators,
+        } = dictcomp;
+
+        self.infer_comprehensions(generators);
+        self.infer_expression(key);
+        self.infer_expression(value);
+    }
+
+    fn infer_set_comprehension_expression_in_scope(&mut self, setcomp: &ast::ExprSetComp) {
+        let ast::ExprSetComp {
+            range: _,
+            elt,
+            generators,
+        } = setcomp;
+
+        self.infer_comprehensions(generators);
+        self.infer_expression(elt);
+    }
+
+    fn infer_comprehensions(&mut self, generators: &[ast::Comprehension]) {
+        let mut generators_iter = generators.iter();
+        let Some(first_generator) = generators_iter.next() else {
+            unreachable!("Expression must contain at least one generator");
+        };
+        self.infer_comprehension(first_generator, true);
+        for generator in generators_iter {
+            self.infer_comprehension(generator, false);
+        }
+    }
+
+    fn infer_comprehension(&mut self, comprehension: &ast::Comprehension, is_first: bool) {
```

Comment:
This method should look up the definition and use `infer_definition_types` query here (thus undering up in `infer_comprehension_definition` to actually do the work), otherwise we'll end up repeating inference of the same comprehension, once in the scope-level query and once in the definition-level query. You can use the existing helper `infer_definition` to do this; see e.g. `infer_function_definition_statement` for an existing case.